### ISSUE TITLE
Mixed precision training

### ISSUE THREAD

**Issue by donghyeon** [tensorpack/tensorpack#1186]
Are there any plans to allow mixed precision training? I have some nvidia gpus which can exploit tensor cores but it seems tensorpack does not support that feature. As far as I know, it needs to reconstruct the model with fp16 and to have master copies of all parameters.

https://nvidia.github.io/OpenSeq2Seq/html/mixed-precision.html
Here's an implementation of a wrapper for tensorflow optimizers to enable this feature, but I doubt that this will work in tensorpack because the model parameters should be defined in fp16 first.

Another workaround for this feature in object detection is here:
https://github.com/tusimple/simpledet
I'm looking forward to that this feature brings huge improvement on speed.

**Comment by ppwwyyxx** [comment]
There is nothing that stops you from doing mixed precision training in tensorpack because the model is implemented by you.

**Comment by ppwwyyxx** [comment]
One example: https://github.com/facebookresearch/ImageNet-Adversarial-Training/blob/326ad41058961d5a871c35fa991082a3f181e169/adv_model.py#L83-L122

**Comment by ppwwyyxx** [comment]
Another tensorpack example that uses fp16 mixed precision training (though on fake data for benchmarking purposes): https://github.com/tensorpack/benchmarks/tree/master/ResNet-MultiGPU#performance--may-2019

**Comment by ArtificialNotImbecile** [comment]
@ppwwyyxx 
I'd like to use loss scaling to prevent underflow together with `Trainer()` api, so I need to implement `optimizer()` method which  can not control how gradients are computed. How can I implement it without modifying the code at `SingleCostTrainer._make_get_grad_fn()`.

**Comment by ppwwyyxx** [comment]
You define the loss in your model, so you can scale it however you want. e.g.: https://github.com/tensorpack/tensorpack/blob/610a8b9baed5ff8f9e8624538141b414f8578d4a/examples/ImageNetModels/imagenet_utils.py#L362-L365

**Comment by ArtificialNotImbecile** [comment]
In optimizers like SGD, I can just downscale my learning rate correspond to reciprocal of loss scale. But some optimizers, say Momentum Optimizer,  combined with some parameter schedule like [fit_one_cycle](https://docs.fast.ai/callbacks.one_cycle.html#What-is-1cycle?), things become subtle how to change parameters from a fp32 precision training schedule.  In theory, I can just transfer this upscale effect to the scale-parameters-of-gradient in back propagation, but it's kind of involved for some optimizers.

**Comment by ppwwyyxx** [comment]
To scale gradients instead of loss, you can use:
https://tensorpack.readthedocs.io/modules/tfutils.html#tensorpack.tfutils.gradproc.ScaleGradient
and for example, https://github.com/tensorpack/tensorpack/blob/c9e03b7356ae1c613707e626452b3e5a0392c86a/examples/SpatialTransformer/mnist-addition.py#L174-L180

**Comment by ArtificialNotImbecile** [comment]
> To scale gradients instead of loss, you can use:
> https://tensorpack.readthedocs.io/modules/tfutils.html#tensorpack.tfutils.gradproc.ScaleGradient
> and for example,
> 
> https://github.com/tensorpack/tensorpack/blob/c9e03b7356ae1c613707e626452b3e5a0392c86a/examples/SpatialTransformer/mnist-addition.py#L174-L180

Thanks!

**Comment by SemanticBeeng** [comment]
@ppwwyyxx Wondering if `transformer/transformer-XL architecture` based models like could be implemented / trained on `tensorpack`.

Kindly advise.

**Comment by ppwwyyxx** [comment]
I think tensorpack supports implementing such models.
